### PR TITLE
fs::safeWriteToFile for Lua & Fix for auth.txt

### LINKED ISSUE #6576: auth.txt Corruption
Both Vanessa and myself have experienced the auth.txt file corrupting (apparently fully deleted with only new entries for that day present). This has been observed on three of Vanessa's worlds and one of mine, all running 0.4.16 (not current dev).

### PR DESCRIPTION
**Pull request by sfan5**
<b>don't</b> squash during merging
(fixes #6576)

### CODE REVIEW

File: doc/lua_api.txt [hunk 1]
```diff
@@ -2380,6 +2380,10 @@ Strings that need to be translated can contain several escapes, preceded by `@`.
       * nil: return all entries,
       * true: return only subdirectory names, or
       * false: return only file names.
+* `minetest.safe_file_write(path, content)`: returns boolean indicating success
+    * Replaces contents of file at path with new contents in a safe (atomic) way.
+      Use this instead of below code when writing e.g. database files:
+      local f = io.open(..., "w"); f:write(newtext); f:close()
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Code example, should be using the inline code notation:
\`local f = io.open(..., "w"); f:write(newtext); f:close()`

File: src/script/lua_api/l_util.cpp [hunk 2]
```diff
@@ -357,6 +357,22 @@ int ModApiUtil::l_get_dir_list(lua_State *L)
 	return 1;
 }
 
+// safe_file_write(path, content)
+int ModApiUtil::l_safe_file_write(lua_State *L)
+{
+	NO_MAP_LOCK_REQUIRED;
+	const char *path = luaL_checkstring(L, 1);
+	const char *content = luaL_checkstring(L, 2);
```
Comment (SmallJoker) [inline src/script/lua_api/l_util.cpp]: Why limit to regular strings only? Modifying binary data should be allowed too IMO.

Comment (sfan5) [reply src/script/lua_api/l_util.cpp]: didn't even consider that, will add

File: doc/lua_api.txt [hunk 3]
```diff
@@ -2380,6 +2380,10 @@ Strings that need to be translated can contain several escapes, preceded by `@`.
       * nil: return all entries,
       * true: return only subdirectory names, or
       * false: return only file names.
+* `minetest.safe_file_write(path, content)`: returns boolean indicating success
+    * Replaces contents of file at path with new contents in a safe (atomic) way.
+      Use this instead of below code when writing e.g. database files:
+      `local f = io.open(..., "w"); f:write(newtext); f:close()`
```
Comment (SmallJoker) [inline doc/lua_api.txt]: How about using the function parameters here for this example?
Also, it's the `"wb"` mode now.

i.e. something like this `local f = io.open(path, "wb"); f:write(content); f:close()` ?

**Comment by sfan5** [comment]
b692454f70cb661a8ce733a4bcbbadcb5edd0049 fc9747eb4b7f75e59a28957bc50f7a78256b3d66

**Pull request closed without merging**